src/GameLogic/NPC/Monster.cs

```diff
@@ -228,22 +228,39 @@ internal async ValueTask RandomMoveAsync()
             return;
         }
 
-        byte randx = (byte)GameLogic.Rand.NextInt(Math.Max(0, this.Position.X - 1), Math.Min(0xFF, this.Position.X + 2));
-        byte randy = (byte)GameLogic.Rand.NextInt(Math.Max(0, this.Position.Y - 1), Math.Min(0xFF, this.Position.Y + 2));
-        if (this.CurrentMap.Terrain.AIgrid[randx, randy] == 1)
+        var moveByMaxX = GameLogic.Rand.NextInt(1, this.Definition.MoveRange + 1);
+        var moveByMaxY = GameLogic.Rand.NextInt(1, this.Definition.MoveRange + 1);
+        byte randx = (byte)GameLogic.Rand.NextInt(Math.Max(0, this.Position.X - moveByMaxX), Math.Min(0xFF, this.Position.X + moveByMaxX + 1));
+        byte randy = (byte)GameLogic.Rand.NextInt(Math.Max(0, this.Position.Y - moveByMaxY), Math.Min(0xFF, this.Position.Y + moveByMaxY + 1));
+
+        switch (this.Definition.ObjectKind)
```

Comment:
Can you move this special treatment to the GuardIntelligence?
Simply add a virtual method like that to the MonsterIntelligenceBase and overwrite it on the GuardIntelligence:
```csharp
public virtual bool CanWalkOn(Point target)
{
    return this.Monster.CurrentMap.Terrain.AIgrid[randx, randy] == 1;
}
```

This way, the code in the `RandomMoveAsync()` looks cleaner and it doesn't have to be extended at this place again when a new kind needs special treatment.